packages/experiments/src/utilities/createSlot.tsx

```diff
@@ -0,0 +1,210 @@
+import * as React from 'react';
+import { memoizeFunction } from 'office-ui-fabric-react';
+
+// TODO: Debug scenarios: is there a way to enable or highlight slots visually? borders? etc.
+// TODO: needs to work with refs / componentRefs/ forwardRefs / etc. Anything else getting bypassed?
+// TODO: make sure no unnecessary attributes are being pasesd to DOM ('as', 'slotAs', etc.)
+
+export type IFactoryComponent<TProps> = React.ReactType<TProps> & {
+  create?: ISlotFactory<TProps>;
+};
+
+export interface IFactoryOptions<TProps> {
+  defaultProp: keyof TProps | 'children';
+}
+
+// Default prop = 'children', default prop type = typeof React children (ReactNode)
+export type IPropsWithChildren<TProps> = TProps & { children?: React.ReactNode };
+
+export type IChildrenProp = React.ReactNode;
+export type INoDefaultProp = never;
+
+/**
+ * An interface definition for defining Slots. Each key in TSlot must point to an IFactoryComponent.
+ */
+export type ISlotDefinition<TSlots> = { [prop in keyof TSlots]: IFactoryComponent<TSlots[prop]> };
+
+/**
+ * Interface for created Slot used internally by components.
+ */
+export type ISlot<TProps> = ((props: IPropsWithChildren<TProps>) => JSX.Element) & { isSlot?: boolean };
+
+/**
+ * Interface for a slot factory that consumes both componnent and user slot prop and generates render output.
+ */
+export type ISlotFactory<TProps> = (componentProps: TProps, userProps: ISlotProp<TProps>) => JSX.Element;
+
+/**
+ * Interface for aggregated Slots objects used internally by components.
+ */
+export type ISlots<TSlots> = { [slot in keyof TSlots]: ISlot<TSlots[slot]> };
+
+export interface IUserProps<TSlots> {
+  classNames: { [prop in keyof TSlots]?: string };
+}
+
+/**
+ * Helper interface components can use for defining Slot properties. This interface defines the following properties:
+ *    1. Component props object (defined by TProps)
+ *    2. ISlotRender function
+ *    3. JSX Elements
+ *    4. Optional shorthand prop, defined by TShorthandProp
+ */
+// TODO: If props object is passed but a required prop is missing, TS coerces to render function and gives a really obscure error.
+//        Is there a way to give a more descriptive error (i.e. "required prop missing")?
+export type ISlotProp<TProps, TShorthandProp = INoDefaultProp> = ISlotRenderFunction<TProps> | JSX.Element | TShorthandProp | TProps;
```

Comment:
can `TShorthandProp` here be simply a string that is typed like this: `ISlotProp<TProps, TShorthandProp keyof TProps>` so you type the getSlots like this: 

```ts
{ icon: ISlotProp<IIconProps, 'iconName'> } 
```

Reply:
Will simplify this even further to be:
```tsx
icon?: IIconSlot;
```
Where the slot types are defined in each subcomponent's types file.